federation/pkg/controller/service/servicecontroller.go

```diff
@@ -0,0 +1,956 @@
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return sc.federationClient.Federation().Clusters().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return sc.federationClient.Federation().Clusters().Watch(options)
+			},
+		},
+		&v1alpha1.Cluster{},
+		controller.NoResyncPeriodFunc(),
+		framework.ResourceEventHandlerFuncs{
+			DeleteFunc: sc.delFromClusterSet,
+			AddFunc:    sc.addToClusterSet,
+		},
+	)
+	return sc
+}
+
+// obj could be an *api.Service, or a DeletionFinalStateUnknown marker item.
+func (sc *ServiceController) enqueueService (obj interface{}) {
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	sc.queue.Add(key)
+}
+
+// Run starts a background goroutine that watches for changes to Federated services
+// and ensures that they have Kubernetes services created, updated or deleted appropriately.
+// federationSyncPeriod controls how often we check the federation's services to
+// ensure that the correct Kubernetes services (and associated DNS entries) exist.
+// This is only necessary to fudge over failed watches.
+// clusterSyncPeriod controls how often we check the federation's underlying clusters and
+// their Kubernetes services to ensure that matching services created independently of the Federation
+// (e.g. directly via the underlying cluster's API) are correctly accounted for.
+
+// It's an error to call Run() more than once for a given ServiceController
+// object.
+func (sc *ServiceController) Run(workers int, stopCh <-chan struct{}) error {
+	if err := sc.init(); err != nil {
+		return err
+	}
+
+	defer runtime.HandleCrash()
+	go sc.serviceController.Run(stopCh)
+	go sc.clusterController.Run(stopCh)
+	for i := 0; i < workers; i++ {
+		go wait.Until(sc.worker, time.Second, stopCh)
+	}
+	go wait.Until(sc.clusterServiceSyncLoop, clusterSyncPeriod, stopCh)
+	<-stopCh
+	glog.Infof("Shutting down Service Controller")
+	sc.queue.ShutDown()
+	return nil
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (sc *ServiceController) worker() {
+	for {
+		func() {
+			key, quit := sc.queue.Get()
+			if quit {
+				return
+			}
+			defer sc.queue.Done(key)
+			err := sc.syncService(key.(string))
+			if err != nil {
+				glog.Errorf("Error syncing service: %v", err)
+			}
+		}()
+	}
+}
+
+func (s *ServiceController) init() error {
+//TODO: waiting for DNS lib
+
+	zones, ok := s.dns.Zones()
+	if !ok {
+		return fmt.Errorf("the cloud provider does not support zone enumeration, which is required for creating load balancers.")
+	}
+	zoneList, err := zones.List()
+	if err != nil {
+		return fmt.Errorf("failed to get zone from cloud provider, will not be able to create load balancers: %v", err)
+	}
+	s.zones = zoneList
```

Comment:
Note: each federation should be configured with a single zone (e.g. "mycompany.com") and we should verify that this zone is one of the zones in zonelist.  And we should not store any of the other zones, in case we accidentally program the wrong zone.